llvm/lib/Transforms/Vectorize/VPlan.h

```diff
@@ -1024,6 +1023,53 @@ class VPInstruction : public VPRecipeWithIRFlags,
   StringRef getName() const { return Name; }
 };
 
+/// A specialization of VPInstruction augmenting it with a dedicated result
+/// type, to be used when the opcode and operands of the VPInstruction don't
+/// directly determine the result type.
+class VPInstructionWithType : public VPInstruction {
```

Comment:
A recipe can "have" a single def, and support the interface of VPValue, by inheriting from it.
A (single def) recipe can have IRflags, by being VPRecipeWithFlags.
A (single def with IRFlags) recipe can have an opcode, by being a VPInstruction.
A (single def with IRFlags with opcode) recipe can have a Type, by being a VPInstructionWithType.
Would using multiple inheritance be better to cope with desired combinations of independent "with's", where IRFlags, opcode, Type, are provided as base classes? In particular, allowing recipes to have Type w/o having IRFlags.

Reply:
Would the Type base class store the Type or just provide an interface for e.g. `getScalarType()`? Is this something that should be explored in a separate PR?

Reply:
I think `VPWidenIntrinsic` also could use it, although it calls `::getScalarType` `::getResultType`. But agreed, I think it can be done as a follow up. It would be nice to get this in to eventually unblock #129508 and #118638

Reply:
It would be nice if we could do something like https://github.com/llvm/llvm-project/commit/d6eb747d784aee6665036cea3c252d77bb94ce60, but unfortunately I've not yet been able to figure out how to make this not actually crash.

Reply:
I managed to sort this out now using a specialization of `CastInfo`, which allows `isa/dyn_cast<VPWithResultType>` to work, effectively turning it into a mixin/trait.

A sketch is here 5c54367f85012baff703d1ad3cd16478d1982fac,  might be worth doing separately>